File: python/354.俄罗斯套娃信封问题.py
#
# @lc app=leetcode.cn id=354 lang=python3
#
# [354] 俄罗斯套娃信封问题
#
from bisect import bisect_left
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        if not envelopes:
            return 0

        res, n = 0, len(envelopes)
        envelopes.sort(key=lambda a:(a[0], -a[1]))
        mem = list()
        for e in envelopes:
            index = bisect_left(mem, e[1])
            if len(mem) == index:
                mem.append(e[1])
            else:
                mem[index] = e[1]
        return len(mem)

    def maxEnvelopesDP(self, envelopes: List[List[int]]) -> int:
        if not envelopes:
            return 0

        res, n = 0, len(envelopes)
        envelopes.sort(key=lambda a:a[0])
        mem = [1]*n
        for i in range(n):
            for j in range(i):
                if envelopes[i][0] > envelopes[j][0] and envelopes[i][1] > envelopes[j][1]:
                    mem[i] = max(mem[j]+1, mem[i])
            res = max(res, mem[i])
        return res

File: cpp/354.俄罗斯套娃信封问题.cpp
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
 if (envelopes.empty()) {
            return 0;
        }
        
        int n = envelopes.size();
        sort(envelopes.begin(), envelopes.end(), [](const auto& e1, const auto& e2) {
            return e1[0] < e2[0] || (e1[0] == e2[0] && e1[1] > e2[1]);
        });

        vector<int> f = {envelopes[0][1]};
        for (int i = 1; i < n; ++i) {
            if (int num = envelopes[i][1]; num > f.back()) {
                f.push_back(num);
            }
            else {
                auto it = lower_bound(f.begin(), f.end(), num);
                *it = num;
            }
        }
        return f.size();
    }
    
};